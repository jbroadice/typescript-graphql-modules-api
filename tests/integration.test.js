const { graphql } = require('graphql');
const schema = require('../src/schema');
const Knex = require('knex');
const knexConfigs = require('../knexfile');

const knexConfig = knexConfigs[process.env.NODE_ENV];

const knex = Knex(knexConfig);


let rootNodeForUser = null;
let runQueryAsUser = null;

describe('User Access Control', () => {
  beforeAll(async () => {
    await knex.migrate.rollback();
    await knex.migrate.latest();
  });
  beforeEach(async () => {
    await knex.seed.run();
    rootNodeForUser = async (username) => {
      const user = await knex('User').select('id').where({ username }).first();
      const context = { userId: user.id, isApiQuery: true };
      return {
        async onQuery(qb) {
          return qb.mergeContext(context);
        },
      };
    };
    runQueryAsUser = async (username, query) => {
      const rootNode = await rootNodeForUser(username);
      return graphql(schema, query, rootNode);
    };
  });

  it('should have our 3 users in the db', async () => {
    const users = await knex('User').select('username', 'password');
    expect(users).toMatchSnapshot();
  });

  describe('Alice', () => {
    it('should be able to fetch users from api', async () => {
      const { data: { users } } = await runQueryAsUser('Alice', '{users { id, username }}');
      expect(users).toMatchSnapshot();
    });

    it('should be not be able to see the passwords of others', async () => {
      const { data: { users } } = await runQueryAsUser('Alice', '{users { id, username, password }}');
      expect(users).toMatchSnapshot();
    });

    it('should be not be able to see the passwords of others', async () => {
      const { data: { users } } = await runQueryAsUser('Alice', '{users { id, username, password }}');
      expect(users).toMatchSnapshot();
    });

    it('should only be able to see their project', async () => {
      const { data: { projects } } = await runQueryAsUser('Alice', '{projects { id, title }}');
      expect(projects).toMatchSnapshot();
    });

    it('should not be able to tunnel through project to see password', async () => {
      const { data: { projects } } = await runQueryAsUser('Alice', '{projects { id, owner { id, username, password }}}');
      expect(projects).toMatchSnapshot();
    });
    it('should not be able to tunnel through user to see project', async () => {
      const { data: { users } } = await runQueryAsUser('Alice', '{users { id, projects { id, title }}}');
      expect(users).toMatchSnapshot();
    });
  });
});
