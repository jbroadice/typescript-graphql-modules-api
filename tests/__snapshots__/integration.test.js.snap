// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Access Control Alice should be able to fetch users from api 1`] = `
Array [
  Object {
    "id": 1,
    "username": "Alice",
  },
  Object {
    "id": 2,
    "username": "Bob",
  },
  Object {
    "id": 3,
    "username": "Eve",
  },
]
`;

exports[`User Access Control Alice should be not be able to see the passwords of others 1`] = `
Array [
  Object {
    "id": 1,
    "password": "alice",
    "username": "Alice",
  },
  Object {
    "id": 2,
    "password": null,
    "username": "Bob",
  },
  Object {
    "id": 3,
    "password": null,
    "username": "Eve",
  },
]
`;

exports[`User Access Control Alice should be not be able to see the passwords of others 2`] = `
Array [
  Object {
    "id": 1,
    "password": "alice",
    "username": "Alice",
  },
  Object {
    "id": 2,
    "password": null,
    "username": "Bob",
  },
  Object {
    "id": 3,
    "password": null,
    "username": "Eve",
  },
]
`;

exports[`User Access Control Alice should not be able to tunnel through project to see password 1`] = `
Array [
  Object {
    "id": 1,
    "owner": Object {
      "id": 1,
      "password": "alice",
      "username": "Alice",
    },
  },
]
`;

exports[`User Access Control Alice should not be able to tunnel through user to see project 1`] = `
Array [
  Object {
    "id": 1,
    "projects": Array [
      Object {
        "id": 1,
        "title": "Project by Alice",
      },
    ],
  },
  Object {
    "id": 2,
    "projects": Array [],
  },
  Object {
    "id": 3,
    "projects": Array [],
  },
]
`;

exports[`User Access Control Alice should only be able to see their project 1`] = `
Array [
  Object {
    "id": 1,
    "title": "Project by Alice",
  },
]
`;

exports[`User Access Control should have our 3 users in the db 1`] = `
Array [
  Object {
    "password": "alice",
    "username": "Alice",
  },
  Object {
    "password": "bob",
    "username": "Bob",
  },
  Object {
    "password": "eve",
    "username": "Eve",
  },
]
`;
